### PR TITLE
Implement PWM audio out for the nrf port

### PR DESCRIPTION
**Pull request by jepler**
This implements PWMAudioOut.  Besides the limitation to 7.5 - 8 bit fidelity, I believe it is now complete.

Testing performed:

My test system is a Particle Xenon with a PAM8302 op-amp https://www.adafruit.com/product/2130 and 8-ohm speaker.  There's no analog filtering between the Xenon's PWM pin and the "A+" input of the amplifier; the "A-" pin is disconnected.  It is powered from VUSB.

I used pin D4, which is *NOT* listed as a low-speed-only pin, but the code does NOT switch the pin to high drive.  This is related to an open issue for general inability to set drive level for pins being used by a "special function" on nrf: #1270

Nothing about the code I've written should limit the usable pins.

All samples I played were 16-bit, generally monophonic at 11025Hz and 22050Hz from the Debian LibreOffice package.

OK  RawSample (8000Hz, 1 channel)
OK      carrier frequency 62.5kHz
OK      Frequency close to 440Hz (measured via o'scope)
OK      Single Play
OK          (too short to do stop/playing test)
OK      Loop
OK          Pause/Resume
OK          Stop/Playing

OK  WaveFile (11025 and 22050Hz, 1 channel)
OK      carrier frequency close to 62.5kHz
OK      Duration of sample close to when played on PC
OK      Single Play
OK          Stop/Playing
OK      Loop
OK          Pause/Resume
OK          Stop/Playing

### CODE REVIEW

**Comment by tannewt** [comment]
I'd be happy to merge this early with bugs but would like it to be in it's own module and class rather than AudioOut. At some point I'd like PWM AudioOut on SAMDs as well for pin flexibility.

**Comment by dhalbert** [comment]
@jepler, there are merge conflicts, so it won't run a build until they're resolved.

**Comment by jepler** [comment]
I've updated the branch, and revised the description of the patch to reflect progress since I filed it.  @tannewt

**Comment by jepler** [comment]
@tannewt While this is unfinished due to pause/resume being missing, I think the renaming and the bulk of the work is done.

File: py/circuitpy_mpconfig.mk [hunk 1]
```diff
@@ -74,8 +74,13 @@ CIRCUITPY_AUDIOIO = $(CIRCUITPY_FULL_BUILD)
...
 CFLAGS += -DCIRCUITPY_AUDIOIO=$(CIRCUITPY_AUDIOIO)
 
+ifndef CIRCUITPY_AUDIOPWMIO
+CIRCUITPY_AUDIOPWMIO = 0
+endif
+CFLAGS += -DCIRCUITPY_AUDIOPWMIO=$(CIRCUITPY_AUDIOPWMIO)
+
 ifndef CIRCUITPY_AUDIOCORE
-CIRCUITPY_AUDIOCORE = $(CIRCUITPY_AUDIOIO)
+CIRCUITPY_AUDIOCORE = $(if $(filter 1, $(CIRCUITPY_AUDIOPWMIO) $(CIRCUITPY_AUDIOBUSIO) $(CIRCUITY_AUDIOIO)),1,0)
```
Comment (tannewt) [inline py/circuitpy_mpconfig.mk]: Is it possible this isn't working when `CIRCUITPY_AUDIOIO = $(CIRCUITPY_FULL_BUILD)` is late bound?

File: shared-bindings/audiopwmio/AudioOut.c [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+    // Properties
+    { MP_ROM_QSTR(MP_QSTR_playing), MP_ROM_PTR(&audiopwmio_audioout_playing_obj) },
+    { MP_ROM_QSTR(MP_QSTR_paused), MP_ROM_PTR(&audiopwmio_audioout_paused_obj) },
+};
+STATIC MP_DEFINE_CONST_DICT(audiopwmio_audioout_locals_dict, audiopwmio_audioout_locals_dict_table);
+
+const mp_obj_type_t audiopwmio_audioout_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_AudioOut,
```
Comment (tannewt) [inline shared-bindings/audiopwmio/AudioOut.c]: Could you rename this to PWMAudioOut or PulseAudioOut too? That way it'll print differently than audioio.AudioOut.

File: py/circuitpy_mpconfig.mk (hunk 1 above)
Comment (jepler) [reply py/circuitpy_mpconfig.mk]: The explanation must be something like that.  I'll dial back the cleverness here.

File: shared-bindings/audiopwmio/AudioOut.c (hunk 2 above)
Comment (tannewt) [reply shared-bindings/audiopwmio/AudioOut.c]: I realize this is super nitpicky but please rename the files and function names too from `audioout` to `pwmaudioout`. Future us will thank you for the consistency.

Comment (jepler) [reply shared-bindings/audiopwmio/AudioOut.c]: Does this latest commit get the renames you were looking for, @tannewt ?

**Comment by jepler** [comment]
@tannewt I think this is ready for another review.  I've updated the initial text of the pull request to reflect the state of testing.

File: ports/nrf/common-hal/audiopwmio/PWMAudioOut.c [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+ *
+ * A PWM sequence can be stopped in its tracks by sending a TASKS_STOP event,
+ * but there's no way to pick up the sequence where it was stopped; you could
+ * start at the start of one of the two sequences, but especially for "single buffer"
+ * sample, this seems undesirable.
+ *
+ * Or, you can stop at the end of a sequence so that you don't duplicate anything
+ * when restarting, but again this is unsatisfactory for a "single buffer" sample.
+ *
+ * For now, I've taken the coward's way and left these methods unimplemented.
```
Comment (tannewt) [inline ports/nrf/common-hal/audiopwmio/PWMAudioOut.c]: You've implemented them! Is this accurate now? One change for looped samples would be that it's desirable to stop on the boundary of the sample.

File: ports/nrf/common-hal/audiopwmio/PWMAudioOut.c [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+ * Perhaps the way forward is to divide even "single buffer" samples into tasks of
+ * only a few ms long, so that they can be stopped/restarted quickly enough that it
+ * feels instant. (This also saves on memory, for long in-memory "single buffer"
+ * samples!)
+ */
+void common_hal_audiopwmio_pwmaudioout_pause(audiopwmio_pwmaudioout_obj_t* self) {
+    self->paused = true;
+    self->pwm->SHORTS = NRF_PWM_SHORT_SEQEND1_STOP_MASK;
+    while(!self->pwm->EVENTS_STOPPED) { /* NOTHING */ }
```
Comment (tannewt) [inline ports/nrf/common-hal/audiopwmio/PWMAudioOut.c]: Does this take a while? We may want to call the background task if it does.

File: ports/nrf/common-hal/audiopwmio/PWMAudioOut.c (hunk 3 above)
Comment (jepler) [reply ports/nrf/common-hal/audiopwmio/PWMAudioOut.c]: Comment updated to reflect the current state of the code.

File: ports/nrf/common-hal/audiopwmio/PWMAudioOut.c (hunk 4 above)
Comment (jepler) [reply ports/nrf/common-hal/audiopwmio/PWMAudioOut.c]: I have resolved this in a way that no longer requires a polling wait.

**Pull request merged**